main/src/library/Iterator.flix

```diff
@@ -216,12 +262,72 @@ namespace Iterator {
     ///
     /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
     ///
-    /// The function `f` must be pure.
+    /// Whether `f` is applied eagerly or lazily depends on its purity:
+    /// - If `f` is pure then it is applied lazily.
+    /// - If `f` is impure then it is applied eagerly (consuming elements of the iterator until and including
+    /// the first element that does *not* satisfy predicate `f`).
     ///
-    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    /// The original iterator `iter` should not be reused.
+    ///
+    pub def takeWhile(f: a -> Bool & ef, iter: Iterator[a]): Iterator[a] & Impure =
+        matchEff f {
+            case Pure   => takeWhileL(f as a -> Bool, iter)
+            case Impure => takeWhileE(f as a -> Bool & Impure, iter)
+        }
+
+    ///
+    /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
+    /// `f` is applied *lazily*.
+    ///
+    /// Does *not* consume any elements of the iterator.
+    ///
+    /// The original iterator `iter` should not be reused.
+    ///
+    def takeWhileL(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure =
+        let Iterator(done, next) = iter;
+        if (done())
+            Iterator(() -> true as & Impure, () -> bug!("Empty iterator") as & Impure)
+        else
+            let x = ref default;
+            let p = () -> {
+                x := next();
+                not f(deref x)
+            };
+            let done1 = () -> done() or p();
+            let next1 = () -> deref x;
+            Iterator(done1, next1)
```

Comment:
So this works by moving every element through the mutable cell `x`. Either that is a brilliant solution to a complicated problem or .. not. Ill have to think about it :) This is essentially your emulation of a push back iterator.

Reply:
Do you feel confident that it works? 

Do you assume that `done` is only called once? I am concerned about that. 

I guess we never specified whether done can be called more than once without changing the iterator.

Reply:
Now that I think about it, there's a case that I missed with this implementation. I will include it as a test.

```flix
    @test
    def takeWhilePure05(): Bool & Impure =
        let iter = (1 :: 2 :: 1 :: Nil): List[Int32] |> Iterator.toIter |> Iterator.takeWhile(x -> x == 1);
        iter |> Iterator.head == Some(1)
            and iter |> Iterator.head == None
            and iter |> Iterator.head == None     // this currently evaluates to Some(1)
```
A solution would be to have a boolean reference denoting whether it has met an element that should not be included already (i.e. that it has "stopped" once already).